File: frontend/src/lib/chql/ast.ts
// ChQL AST — mirrors Rust backend's serde-serialized shape exactly.

export type CmpOp =
  | 'eq'
  | 'neq'
  | 'gt'
  | 'gte'
  | 'lt'
  | 'lte'
  | 'ilike_contains'
  | 'regex'
  | 'wildcard';

export type Value =
  | { kind: 'str'; v: string }
  | { kind: 'num'; v: number }
  | { kind: 'date'; v: string };

export type Ast =
  | { type: 'and'; left: Ast; right: Ast }
  | { type: 'or'; left: Ast; right: Ast }
  | { type: 'not'; expr: Ast }
  | { type: 'cmp'; field: string; op: CmpOp; value: Value }
  | { type: 'in_range'; field: string; lo: Value; hi: Value };

export const VALID_FIELDS = [
  'state',
  'branch',
  'repo',
  'repo_id',
  'stage',
  'exit_code',
  'created_at',
  'completed_at',
] as const;

export type ValidField = (typeof VALID_FIELDS)[number];

/** Fields where value is a date (ISO-8601 string). */
export const DATE_FIELDS: ReadonlySet<string> = new Set(['created_at', 'completed_at']);

/** Fields where value is numeric. */
export const NUMERIC_FIELDS: ReadonlySet<string> = new Set(['exit_code']);

/** Regex for ISO-8601 date strings (YYYY-MM-DD). */
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

export function isDateString(s: string): boolean {
  return DATE_RE.test(s);
}

File: crates/ci-controller/src/query/ast.rs
//! ChQL AST types. Serialization is the shared contract with the TS parser —
//! the JSON shape must exactly match `tests/corpus/*.json`. See
//! `local/plans/CHQL.md` for the grammar.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Ast {
    And {
        left: Box<Ast>,
        right: Box<Ast>,
    },
    Or {
        left: Box<Ast>,
        right: Box<Ast>,
    },
    Not {
        expr: Box<Ast>,
    },
    Cmp {
        field: String,
        op: CmpOp,
        value: Value,
    },
    InRange {
        field: String,
        lo: Value,
        hi: Value,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CmpOp {
    Eq,
    Neq,
    Gt,
    Gte,
    Lt,
    Lte,
    IlikeContains,
    Regex,
    Wildcard,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Value {
    Str { v: String },
    Num { v: f64 },
    Date { v: String },
}
